### DISCUSSION TITLE
ZipReader:filenameEncoding:function

### DISCUSSION THREAD

**Discussion by nenge123** [gildas-lormeau/zip.js#507] [General]
the config can be set a function? replace `decodeText`
`filenameEncoding:string` replace to  `new TextDecoder(str).decode` => `filenameEncoding(uint8,encoding)`

`filenameEncoding:function` = `filenameEncoding(uint8,encoding)`

**Comment by nenge123** [comment]
or  export method
open `zip.decodeText`,  we can edit  the `decodeText` to auto decode function.

because zip.js have more cdn link now

**Comment by gildas-lormeau** [comment]
Thank you for the suggestion. I agree that these options could be useful. It's implemented in the version [`2.7.42`](https://github.com/gildas-lormeau/zip.js/releases/tag/v2.7.42).

**Reply by nenge123** [reply]
`decode(rawFilename, filenameUTF8 ? CHARSET_UTF8 : filenameEncoding || CHARSET_CP437)`

=>

`decode(rawFilename, filenameUTF8 ? CHARSET_UTF8 : filenameEncoding || CHARSET_CP437) || decodeText(rawFilename,CHARSET_CP437)`

if `options.decodeText`  nothing return ,will no file return 
or some one need the 'cp437' decode;
```javascript
getEntries({
    filenameEncoding:'utf-8',
    decodeText(buf,encoding){
        let text = new TextDecoder().decode(buf);
        let newbuf = new TextEncoder().encode(text);
        console.log(encoding,buf,text,newbuf.byteLength);
        return newbuf.byteLength>buf.byteLength ? new TextDecoder('gbk').decode(buf):text;
    }
});
//file is not utf8
//console.log(encoding,buf,newbuf.byteLength);
//utf-8 ,Uint8Array(37),69
//cp437 ,Uint8Array [buffer: ArrayBuffer(366)],0
```

**Reply by gildas-lormeau** [reply]
I'm sorry but I can't understand the problem. In what case could the string be empty? Do you have a specific example of a file name that would cause this problem?

**Comment by nenge123** [comment]
@gildas-lormeau 
when  user `decodeText` have `new TextDecoder(encoding).decode(buf)` 

```
getEntries({
    filenameEncoding:'utf-8',
    decodeText(buf,encoding){
        return new TextDecoder(encoding).decode(buf);
    }
});
```

**Reply by gildas-lormeau** [reply]
But `buf` is not empty, is it?

**Reply by nenge123** [reply]
you can try any file.
`encoding !== utf-8  ` the comment is `cp437`

**Comment by nenge123** [comment]
@gildas-lormeau 
```javascript
new TextDecoder('cp437').decode(new Uint8Array([12]))
//Uncaught RangeError: Failed to construct 'TextDecoder': The encoding label provided ('cp437') is invalid.
    at <anonymous>:1:1
```

**Reply by gildas-lormeau** [reply]
You have to use this code (or another third-party library) to decode strings encoded in CP437: https://github.com/gildas-lormeau/zip.js/blob/master/lib/core/util/cp437-decode.js. See https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/encoding

**Comment by gildas-lormeau** [comment]
Having written the previous answer and after some thought, I can understand that dealing with `CP437` encoding on your end is a pain. I've implemented your proposal. If the function `encodeText` or `decodeText` returns `undefined` then the corresponding default function in zip.js will be called as a fallback function.

I should publish a new version in the coming days.  Feel free to test the implementation on the `master` branch meanwhile.
